### PR TITLE
Improve the random security

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/1144

### CODE REVIEW

File: neo/Network/RPC/TransactionManager.cs [hunk 1]
```diff
@@ -50,11 +49,17 @@ public TransactionManager(RpcClient rpc, UInt160 sender)
         /// <returns></returns>
         public TransactionManager MakeTransaction(byte[] script, TransactionAttribute[] attributes = null, Cosigner[] cosigners = null, long networkFee = 0)
         {
+            byte[] buffer = new byte[sizeof(uint)];
+            using (var random = System.Security.Cryptography.RandomNumberGenerator.Create())
```
Comment (vncoelho) [inline neo/Network/RPC/TransactionManager.cs]: @shargon, maybe create a generic function `getBytes` for avoiding repeating this `using`.
In addition, the function can size the `sizeof`.

File: neo/Wallets/Wallet.cs [hunk 2]
```diff
@@ -299,49 +299,55 @@ public Transaction MakeTransaction(byte[] script, UInt160 sender = null, Transac
...
                     }
-                }
 
-                UInt160[] hashes = tx.GetScriptHashesForVerifying(snapshot);
+                    UInt160[] hashes = tx.GetScriptHashesForVerifying(snapshot);
 
-                // base size for transaction: includes const_header + attributes + cosigners with scopes + script + hashes
-                int size = Transaction.HeaderSize + attributes.GetVarSize() + cosigners.GetVarSize() + script.GetVarSize() + IO.Helper.GetVarSize(hashes.Length);
+                    // base size for transaction: includes const_header + attributes + cosigners with scopes + script + hashes
+                    int size = Transaction.HeaderSize + attributes.GetVarSize() + cosigners.GetVarSize() + script.GetVarSize() + IO.Helper.GetVarSize(hashes.Length);
```
Comment (vncoelho) [inline neo/Wallets/Wallet.cs]: why this changed? and below?

File: neo/Network/RPC/TransactionManager.cs (hunk 1 above)
Comment (vncoelho) [reply neo/Network/RPC/TransactionManager.cs]: I just saw that for `wallet` it is different.

File: neo/Wallets/Wallet.cs (hunk 2 above)
Comment (shargon) [reply neo/Wallets/Wallet.cs]: because now is inside the using

**Comment by erikzhang** [comment]
I don't think it would be a good idea to replace all `Random`s with `RandomNumberGenerator`. These random numbers don't need too much security, so we should choose a better performance `Random`.

**Comment by shargon** [comment]
I will change it to maintain the `Random` class for performance, but without static variable

File: neo/Network/RPC/TransactionManager.cs [hunk 3]
```diff
@@ -50,11 +49,12 @@ public TransactionManager(RpcClient rpc, UInt160 sender)
         /// <returns></returns>
         public TransactionManager MakeTransaction(byte[] script, TransactionAttribute[] attributes = null, Cosigner[] cosigners = null, long networkFee = 0)
         {
+            var random = new Random();
```
Comment (vncoelho) [inline neo/Network/RPC/TransactionManager.cs]: It it good?
Why not keep as before, initializing and let the seed feed the generator?

Comment (shargon) [reply neo/Network/RPC/TransactionManager.cs]: It's safer

Comment (vncoelho) [reply neo/Network/RPC/TransactionManager.cs]: Can we predict next Random based on historical values? @igormcoelho

File: neo/Consensus/ConsensusContext.cs [hunk 4]
```diff
@@ -274,6 +273,7 @@ internal void EnsureMaxBlockSize(IEnumerable<Transaction> txs)
 
         public ConsensusPayload MakePrepareRequest()
         {
+            var random = new Random();
```
Comment (erikzhang) [inline neo/Consensus/ConsensusContext.cs]: I don't think that `MakePrepareRequest()` will be accessed by multiple threads at the same time. There should be no risk.

File: neo/Network/RPC/TransactionManager.cs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿using Neo.Cryptography.ECC;
+using Neo.Cryptography.ECC;
```
Comment (erikzhang) [inline neo/Network/RPC/TransactionManager.cs]: Why this is changed but didn't cause travis to fail?

File: neo/Consensus/ConsensusContext.cs (hunk 4 above)
Comment (shargon) [reply neo/Consensus/ConsensusContext.cs]: I don't think so too, but always is better to use local, maybe in the future other developer use it and we forget this, i prefer to keep it local in order to be safe.

File: neo/Network/RPC/TransactionManager.cs (hunk 5 above)
Comment (shargon) [reply neo/Network/RPC/TransactionManager.cs]: Good question

Comment (shargon) [reply neo/Network/RPC/TransactionManager.cs]: The file was with BOM and here it is removed, I don't know why it passed previously, but here is good.

File: neo/Consensus/ConsensusContext.cs (hunk 4 above)
Comment (shargon) [reply neo/Consensus/ConsensusContext.cs]: We have alternatives. 

1- Local variable as now
2- Static variable (maybe in Helper.cs) using `ThreadStaticAttribute`
```csharp
        [ThreadStatic]
        private static Random _rand;

        /// <summary>
        /// Thread safe Random
        /// </summary>
        private static Random Rand
        {
            get
            {
                if (_rand == null) _rand = new Random();
                return _rand;
            }
        }
```

Comment (erikzhang) [reply neo/Consensus/ConsensusContext.cs]: `ThreadStaticAttribute` will create different values for each thread. It is not a good option.

We only need one `Random`. Although there may be multiple threads accessing it, different threads are not accessed at the same time, so it is safe.

Comment (shargon) [reply neo/Consensus/ConsensusContext.cs]: > so it is safe.

It's safe now, but dangerous, I preffer to use a local variable, the speed will be almost the same.

**Pull request merged**